File: spicedb-java/lib/src/main/java/com/authzed/spicedb/LookupOptions.java
package com.authzed.spicedb;

/**
 * Immutable call-level options for the lookup operations.
 *
 * <p>This type exists so that a new lookup option is a new component here rather than a new
 * overload. {@code with_debug} arrived upstream with nowhere to go and became a second {@code
 * lookupResources} overload; the option after it would have needed a third. See root DESIGN.md,
 * "RULE: Every RPC wrapper must have one place to add an option".
 *
 * @param debug asks the server to attach debug information to the error raised when the lookup
 *     fails by exceeding the maximum dispatch depth. It has no effect on a successful call, and
 *     none on any other failure.
 */
public record LookupOptions(boolean debug) {

  /** Options with nothing set: the client's defaults apply. */
  public static LookupOptions none() {
    return new LookupOptions(false);
  }

  /** Returns a copy asking the server for debug information on a depth failure. */
  public LookupOptions withDebug(boolean debug) {
    return new LookupOptions(debug);
  }
}

File: spicedb-csharp/SpiceDB.Client/LookupOptions.cs
namespace SpiceDB.Client;

/// <summary>
/// Call-level options for the lookup operations:
/// <see cref="SpiceDBClient.LookupResourcesWithOptionsAsync"/> and
/// <see cref="SpiceDBClient.LookupSubjectsWithOptionsAsync"/>.
/// </summary>
/// <remarks>
/// This type exists so that a new lookup option is a new property here rather
/// than a new parameter. <c>with_debug</c> arrived upstream and, with nowhere
/// to put it, became a required positional parameter on
/// <c>LookupResourcesAsync</c> — breaking every existing caller for the sake
/// of one optional field. See root DESIGN.md, "RULE: Every RPC wrapper must
/// have one place to add an option".
/// </remarks>
public sealed class LookupOptions
{
    /// <summary>
    /// Asks the server to attach debug information to the error when the
    /// lookup fails by exceeding the maximum dispatch depth. It has no effect
    /// on a successful call, and no effect on any other failure.
    /// </summary>
    public bool WithDebug { get; init; }
}
